baggage/baggage.go

```diff
@@ -61,58 +61,75 @@ type Property struct {
 	// hasValue indicates if a zero-value value means the property does not
 	// have a value or if it was the zero-value.
 	hasValue bool
+
+	// hasData indicates whether the created property contains data or not.
+	// Properties that do not contain data are invalid with no other check
+	// required.
+	hasData bool
 }
 
 func NewKeyProperty(key string) (Property, error) {
-	p := Property{}
 	if !keyRe.MatchString(key) {
-		return p, fmt.Errorf("%w: %q", errInvalidKey, key)
+		return newInvalidProperty(), fmt.Errorf("%w: %q", errInvalidKey, key)
 	}
-	p.key = key
+
+	p := Property{key: key, hasData: true}
 	return p, nil
 }
 
 func NewKeyValueProperty(key, value string) (Property, error) {
-	p := Property{}
 	if !keyRe.MatchString(key) {
-		return p, fmt.Errorf("%w: %q", errInvalidKey, key)
+		return newInvalidProperty(), fmt.Errorf("%w: %q", errInvalidKey, key)
 	}
 	if !valueRe.MatchString(value) {
-		return p, fmt.Errorf("%w: %q", errInvalidValue, value)
+		return newInvalidProperty(), fmt.Errorf("%w: %q", errInvalidValue, value)
+	}
+
+	p := Property{
+		key:      key,
+		value:    value,
+		hasValue: true,
+		hasData:  true,
 	}
-	p.key = key
-	p.value = value
-	p.hasValue = true
 	return p, nil
 }
 
+func newInvalidProperty() Property {
+	return Property{}
+}
+
 // parseProperty attempts to decode a Property from the passed string. It
 // returns an error if the input is invalid according to the W3C Baggage
 // specification.
 func parseProperty(property string) (Property, error) {
-	p := Property{}
 	if property == "" {
-		return p, nil
+		return newInvalidProperty(), nil
 	}
 
 	match := propertyRe.FindStringSubmatch(property)
 	if len(match) != 4 {
-		return p, fmt.Errorf("%w: %q", errInvalidProperty, property)
+		return newInvalidProperty(), fmt.Errorf("%w: %q", errInvalidProperty, property)
 	}
 
+	p := Property{hasData: true}
 	if match[1] != "" {
 		p.key = match[1]
 	} else {
 		p.key = match[2]
 		p.value = match[3]
 		p.hasValue = true
 	}
+
 	return p, nil
 }
 
 // validate ensures p conforms to the W3C Baggage specification, returning an
 // error otherwise.
 func (p Property) validate() error {
+	if !p.hasData {
+		return errInvalidProperty
```

Comment:
Specifically this is identifying an invalid key. Can we update this returned error to match what is done below to make sure our returned errors are as actionable as possible? i.e.

```go
return errFunc(fmt.Errorf("%w: %q", errInvalidKey, p.key))
```

Reply:
Sure, I've done so. I've kept it an `errInvalidProperty` though, as technically, the reason for having an invalid property could have been something different than the key (the actual reason would have been provided when the invalid property was initially provided though).